hbase-server/src/main/java/org/apache/hadoop/hbase/master/startupprogress/MasterStartupProgress.java

```diff
@@ -0,0 +1,83 @@
+/**
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hbase.master.startupprogress;
+
+import static org.apache.hadoop.hbase.util.EnvironmentEdgeManager.currentTime;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import org.apache.hadoop.hbase.master.HMaster;
+import org.apache.yetus.audience.InterfaceAudience;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ *  MasterStartupProgress tracks all the different phases in HMaster startup process.
+ *  Once the master startup is complete, you can't begin or end phase.
+ *  For read access, call {@link #createView()} to create a consistent view with
+ *  a clone of the data.
+ */
+@InterfaceAudience.Private
+public class MasterStartupProgress {
```

Comment:
We have MonitoredTask and MonitoredTaskImpl which appear to do very similar things to what you're doing here.

Did you look at MonitoredTask and is there a reason we cannot use/extend it instead of writing net-new code?